### PR TITLE
[OIS] Add TF-M support to Matter examples

### PR DESCRIPTION
**Pull request by ATmobica**
Enhancement to Open IoT SDK platform which demonstrates the usage of TF-M component in Matter examples.

 - Improving OIS platform storage - storage library with custom Flash Block Device class.
 - Add TF-M support in Matter examples (shell and lock-app)
 - TF-M PSA protected storage support in Matter example (lock-app)

### CODE REVIEW

**Comment by ATmobica** [comment]
@Damian-Nordic I'm sorry but there is some unexpected issue with the previous PR #23939. This is the new one with the added description of TF-M in docs.

**Comment by Damian-Nordic** [comment]
> @Damian-Nordic I'm sorry but there is some unexpected issue with the previous PR #23939. This is the new one with the added description of TF-M in docs.

Heh, I don't understand why github claims that I merged 0 commits from #23939. I haven't done anything...

**Comment by ATmobica** [comment]
@Damian-Nordic No, not me, but @adbridge is aware of that. So she probably will be interested in your PR :)

File: scripts/examples/openiotsdk_example.sh [hunk 1]
```diff
@@ -113,6 +113,10 @@ function build_with_cmake() {
     PATH=$(echo "$PATH" | sed 's/:/\n/g' | grep -v "$PW_ARM_CIPD_INSTALL_DIR" | xargs | tr ' ' ':')
 
     cmake -G Ninja -S "$EXAMPLE_PATH" -B "$BUILD_PATH" --toolchain="$TOOLCHAIN_PATH" "${BUILD_OPTIONS[@]}"
+    if [ -f "$BUILD_PATH"/_deps/tf-m-src/tools/requirements.txt ]; then
```
Comment (andy31415) [inline scripts/examples/openiotsdk_example.sh:116]: What does this do (i.e. why is this needed)?

If this shell script changes, should there be a change in `scripts/build/builders/openiotsdk.py` ?

Comment (ATmobica) [reply scripts/examples/openiotsdk_example.sh:118]: We need to install the required Python packages for TF-M tools which are used during its component build. I added some comments.
Yes, definitely, I added changes there.

File: scripts/build/builders/openiotsdk.py [hunk 2]
```diff
@@ -58,6 +59,10 @@ def ExamplePath(self):
         return os.path.join(self.root, 'examples', self.app.ExampleName, 'openiotsdk')
 
     def generate(self):
+        # Remove access to ARM GCC toolchain from Matter environment, use higher version from OIS environment
+        cmd = "PATH=$(echo $PATH | sed 's/:/\n/g' | grep -v $PW_ARM_CIPD_INSTALL_DIR | xargs | tr ' ' ':')"
+        self._Execute(['bash', '-c', cmd],
```
Comment (andy31415) [inline scripts/build/builders/openiotsdk.py:64]: if you execute `bash -c` would the path changes not only persist in the current shell invocation (i.e. CMAKE will not see it)?

You may need to execute the entire PATH + cmake into one large bash command ... I believe this is what we do for idf.py builds.

Also wondering if this path change is needed during the build phase or if cmake caches compiler paths.

File: scripts/build/builders/openiotsdk.py [hunk 3]
```diff
@@ -72,6 +77,12 @@ def _build(self):
         self._Execute(['bash', '-c', cmd],
```
Comment (andy31415) [inline scripts/build/builders/openiotsdk.py:77]: This was already here, but wondering: why do we delete outputs to force re-linking. The comment says "force linking" ... but why?  this means that this does a `ld` on every build regardless if I have changes or not ... that seems wrong.

File: scripts/build/builders/openiotsdk.py [hunk 4]
```diff
@@ -72,6 +77,12 @@ def _build(self):
...
                       title='Remove old artifacts ' + self.identifier)
 
+        # Install the required Python packages for TF-M tools used during its components build
+        tfm_requirements_path = os.path.join(self.output_dir, "_deps", "tf-m-src", "tools", "requirements.txt")
+        if os.path.exists(tfm_requirements_path):
+            self._Execute(['pip3', 'install', '-r', tfm_requirements_path],
```
Comment (andy31415) [inline scripts/build/builders/openiotsdk.py:83]: This will change the global build environment for the user ... so sideffects outside the build directory for the user.

What are these requirements? Could they be installed as part of general chip requirements?

Comment (ATmobica) [reply scripts/build/builders/openiotsdk.py:83]: Agree, we can use Python virtual env in this case.
These are the required package for the current version:
```
cbor2
click
cryptography
pyasn1
imgtool>=1.6.0
Jinja2>=2.10.3
PyYAML
pyhsslms
ecdsa
kconfiglib
```
But it could be different depending on the TF-M version, so I preferred to install it dynamically during the build process.

File: scripts/build/builders/openiotsdk.py (hunk 2 above)
Comment (ATmobica) [reply scripts/build/builders/openiotsdk.py:64]: I just noticed that it won't be needed anymore because the Pigweed environment now has the correct version of GCC ARM compiler (10.3-2021.10).

File: scripts/build/testdata/dry_run_openiotsdk-shell.txt [hunk 5]
```diff
@@ -1,6 +1,10 @@
 # Commands will be run in CHIP project root.
 cd "{root}"
 
+# Prepare environment openiotsdk-shell
+bash -c 'PATH=$(echo $PATH | sed '"'"'s/:/
+/g'"'"' | grep -v $PW_ARM_CIPD_INSTALL_DIR | xargs | tr '"'"' '"'"' '"'"':'"'"')'
```
[the remark this answers is not in the document]
Comment (ATmobica) [reply scripts/build/testdata/dry_run_openiotsdk-shell.txt:6]: Outdated

File: scripts/build/builders/openiotsdk.py (hunk 4 above)
Comment (ATmobica) [reply scripts/build/builders/openiotsdk.py:83]: I checked the required packages and only imgtool was missed in general chip requirements. I'll at it and it should work without the additional steps during the build process.

File: scripts/build/builders/openiotsdk.py (hunk 3 above)
Comment (ATmobica) [reply scripts/build/builders/openiotsdk.py:77]: Remnant of the first implementation, I will remove it.

**Comment by ATmobica** [comment]
@andy31415 Can I ask you for a re-review?

File: src/app/chip_data_model.cmake [hunk 6]
```diff
@@ -98,7 +105,8 @@ function(chip_configure_data_model APP_TARGET)
           OUTPUT_FILES  APP_GEN_FILES
         )
 
-        target_include_directories(${APP_TARGET} PRIVATE "${APP_GEN_DIR}")
+        message(STATUS "SCOPE: ${SCOPE}")
```
Comment (Damian-Nordic) [inline src/app/chip_data_model.cmake:108]: Debug message

**Review by Damian-Nordic — APPROVED** [review]

**Comment by ATmobica** [comment]
@andy31415 @bzbarsky-apple Can we go forward with this PR?

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
> @andy31415 @bzbarsky-apple Can we go forward with this PR?

@ATmobica some CI was failing, restarted them. However I will bypass pullapprove for merging purposes as I cannot kick that one.

**Review by bzbarsky-apple — APPROVED** [review]
